byzcoin/service.go

```diff
@@ -1001,106 +1001,115 @@ func (s *Service) downloadDB(sb *skipchain.SkipBlock) error {
 			}
-			if sb.Index != st.GetIndex() {
-				log.Lvl2("Downloading corresponding block")
-				skCl := skipchain.NewClient()
-				// TODO: add a client API to fetch a specific block and its proof
-				search, err := skCl.GetSingleBlockByIndex(roster, sb.SkipChainID(), st.GetIndex())
-				if err != nil {
-					return errors.New("couldn't get correct block for verification: " + err.Error())
-				}
-				sb = search.SkipBlock
+			if db == nil {
+				db, bucketName = s.GetAdditionalBucket([]byte(idStr))
+				nonce = resp.Nonce
 			}
-			var header DataHeader
-			err = protobuf.Decode(sb.Data, &header)
+			// And store all entries in our local database.
+			err = db.Update(func(tx *bbolt.Tx) error {
+				bucket := tx.Bucket(bucketName)
+				for _, kv := range resp.KeyValues {
+					err := bucket.Put(kv.Key, kv.Value)
+					if err != nil {
+						return err
+					}
+				}
+				return nil
+			})
 			if err != nil {
-				return errors.New("couldn't unmarshal header: " + err.Error())
+				return fmt.Errorf("Couldn't store entries: %s", err.Error())
 			}
-			if !bytes.Equal(st.GetRoot(), header.TrieRoot) {
-				return errors.New("got wrong database, merkle roots don't work out")
+			if len(resp.KeyValues) < catchupFetchDBEntries {
+				break
 			}
+		}
 
-			// Finally initialize the stateTrie using the new database.
-			s.stateTriesLock.Lock()
-			s.stateTries[idStr] = st
-			s.stateTriesLock.Unlock()
-			log.Lvlf1("%s: successfully downloaded database for chain %s", s.ServerIdentity(),
-				idStr)
-			return nil
-		}()
-		if err == nil {
-			return nil
+		// Check the new trie is correct
+		st, err := loadStateTrie(db, bucketName)
+		if err != nil {
+			return errors.New("couldn't load state trie: " + err.Error())
+		}
+		skCl := skipchain.NewClient()
+		// TODO: avoid contacting ourselves
+		if sb.Index != st.GetIndex() {
+			log.Lvl2("Downloading corresponding block", sb.Index, st.GetIndex())
+			// TODO: add a client API to fetch a specific block and its proof
+			search, err := skCl.GetSingleBlockByIndex(newRoster, sb.SkipChainID(), st.GetIndex())
+			if err != nil {
+				return errors.New("couldn't get correct block for verification: " + err.Error())
+			}
+			sb = search.SkipBlock
+		}
+		var header DataHeader
+		err = protobuf.Decode(sb.Data, &header)
+		if err != nil {
+			return errors.New("couldn't unmarshal header: " + err.Error())
+		}
+		if !bytes.Equal(st.GetRoot(), header.TrieRoot) {
+			return errors.New("got wrong database, merkle roots don't work out")
+		}
+
+		// Finally initialize the stateTrie using the new database.
+		s.stateTriesLock.Lock()
+		s.stateTries[idStr] = st
+		s.stateTriesLock.Unlock()
+		chain, err := skCl.GetUpdateChain(newRoster, sb.SkipChainID())
+		if err != nil {
+			return err
 		}
-		log.Errorf("Couldn't load database from %s - got error %s", roster.List[0], err)
+		for _, sb := range chain.Update {
+			log.Lvlf2("Storing block %d: %x", sb.Index, sb.CalculateHash())
+			s.db().Store(sb)
```

Comment:
when you're storing the skipblock, can you also check that you're only storing the skipblocks that are at the same index or below the stateTrie? otherwise the blocks and the trie might become out of sync

Reply:
for every block that is stored in the db, `updateTrieCallback` will be called. If the block is newer than the trie, which might happen if another block gets created while the db is downloaded, `updateTrieCallback` will make sure that the block will be added to the trie:

- If it's just one new block, all is well, as this is the normal path in `updateTrieCallback`
- if it's two new blocks, the go-routine with `s.catchUp` in it will be called, and all is well
- if it's three or more new blocks, multiple go-routines will be called, each with a `s.catchUp` call. The first `s.catchUp` will update to the latest block, and all further `s.catchUp` will pass through as there is nothing to update to

And no, I didn't think this through while writing the code ;)